### PR TITLE
Added GitHub workflows [PCL-2731]

### PR DESCRIPTION
**Pull request by gnawme**
This PR is a companion to #4629; per @kunaltyagi, it only adds the GitHub Workflow to enable `clang-tidy` checks in CI, but does not actually enable `clang-tidy` in the PCL repo itself.

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Any thoughts on why we don't see Github workflow in the list of checks?

File: .github/workflows/clang-tidy-modern.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+run-clang-tidy-10.py -config='' -header-filter='^((?!.*kiss_fft\.h.*).)*$,.*' -fix
```
Comment (kunaltyagi) [inline .github/workflows/clang-tidy-modern.sh:2]: Please remove the filter for kiss_fft. Let that come in with the appropriate checks

**Review by kunaltyagi — COMMENTED** [review]
Apart from lack of github workflow viz, and minor comment, LGTM

Thoughts on adding a docker image for clang-tidy? (env + clang + git)? @larshg @mvieth

**Comment by larshg** [comment]
> Apart from lack of github workflow viz, and minor comment, LGTM
> 
> Thoughts on adding a docker image for clang-tidy? (env + clang + git)? @larshg @mvieth

Maybe just run it directly on ubuntu_latest / fixed ubuntu.
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md

Then its just clang-tidy that needs installed?

It does give issues, that MS might change the versions installed, but it would avoid pulling a docker image?

**Comment by kunaltyagi** [comment]
IIRC, doesn't clang-tidy require all dependencies to be installed?

**Comment by larshg** [comment]
> IIRC, doesn't clang-tidy require all dependencies to be installed?

Ahh, that might be. I'm not that familiar with clang/tidy.

I can see Cmake is run as well, which will fail with missing dependencies.

Is git required explicitly installed?

**Comment by gnawme** [comment]
> Any thoughts on why we don't see Github workflow in the list of checks?

Not sure, it was triggered automatically in the previous PR once I added the workflow files

File: .github/workflows/tidy-modern.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+name: tidy-modern
+
+on:
+  push:
+  pull_request:
```
Comment (kunaltyagi) [inline .github/workflows/tidy-modern.yml:5]: Could this be the culprit? Empty push and pull triggers

Comment (gnawme) [reply .github/workflows/tidy-modern.yml:5]: Oddly, I never saw a `clang-tidy` build step in #4560, although the GitHub workflow plainly ran there.

File: compile_commands.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[
```
Comment (kunaltyagi) [inline compile_commands.json:1]: This files doesn't belong in the PR

**Review by kunaltyagi — CHANGES_REQUESTED** [review]

**Comment by kunaltyagi** [comment]
I think we need to merge this into master before the github action is activated. We saw something similar with Azure pipelines as well

File: compile_commands.json (hunk 3 above)
Comment (gnawme) [reply compile_commands.json:1]: I included it so the GitHub action wouldn't error out looking for it.

File: .github/workflows/tidy-modern.yml (hunk 2 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: Apparently, they changed the syntax; it's now `on: [push, pull_request]`

Once I did that, it started triggering.

**Comment by gnawme** [comment]
> I think we need to merge this into master before the github action is activated. We saw something similar with Azure pipelines as well

See reply to your previous comment.

File: compile_commands.json (hunk 3 above)
Comment (kunaltyagi) [reply compile_commands.json:1]: In that case, the file should be generated by the CMakeLists.txt

Comment (gnawme) [reply compile_commands.json:1]: That'll happen once this is merged with #4629. I just wanted to see this PR pass all CI checks.

I can remove `compile_commands.json` now that has happened.

**Comment by mvieth** [comment]
Closing and re-opening to trigger checks ...

File: .github/workflows/clang-tidy-modern.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+run-clang-tidy-10.py -config='' -header-filter='.*' -fix
```
Comment (mvieth) [inline .github/workflows/clang-tidy-modern.sh:2]: What's your reason to move this to its own file? As long as it is just one line, I would leave it in tidy-modern.yml

File: .github/workflows/tidy-modern.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+name: tidy-modern
```
Comment (mvieth) [inline .github/workflows/tidy-modern.yml:1]: How about renaming this (the yml-file as well) to clang-tidy? "modern" implies that this runs only the "modernize" checks, but in the future we probably also want to run other checks from clang-tidy

File: .github/workflows/tidy-modern.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+on: [push, pull_request]
+
+jobs:
+  tidy:
+    runs-on: ubuntu-latest
+    container:
+      image: pointcloudlibrary/env:18.04
```
Comment (mvieth) [inline .github/workflows/tidy-modern.yml:9]: I would suggest that we update this to 20.04 or even 22.04

File: .github/workflows/tidy-modern.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    
+    steps:
+    - name: Install git
```
Comment (mvieth) [inline .github/workflows/tidy-modern.yml:12]: The `pointcloudlibrary/env` docker already contains git, so I don't think this is necessary?

File: .github/workflows/tidy-modern.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+          chmod +x llvm.sh
+          ./llvm.sh 10
+          apt-get update
+          apt-get install -y clang-tidy-10 clang-tools-10 python-pip
+          pip install pyyaml
+          cmake .
+          ./.github/workflows/clang-tidy-modern.sh
+          git diff > tidy.patch
+          cat tidy.patch
+          [ ! -s formatting.patch ]
```
Comment (mvieth) [inline .github/workflows/tidy-modern.yml:33]: Should this be `tidy.patch`?

**Review by mvieth — COMMENTED** [review]
Please also rebase on the current master branch, otherwise the `compile_commands.json` is not generated

File: .github/workflows/clang-tidy-modern.sh (hunk 4 above)
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh]: Done

File: .github/workflows/tidy-modern.yml (hunk 6 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: 22.04 it is

File: .github/workflows/tidy-modern.yml (hunk 8 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: Seems so...

**Comment by gnawme** [comment]
@mvieth Opened #5304 to address issues that have crept in since #4629

File: .github/workflows/clang-tidy.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+name: tidy-modern
```
Comment (mvieth) [inline .github/workflows/clang-tidy.yml:1]:
```suggestion
name: clang-tidy
```

File: .github/workflows/clang-tidy.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    - name: Run clang-tidy
+      run: |
+          wget https://apt.llvm.org/llvm.sh
+          chmod +x llvm.sh
+          ./llvm.sh 14
+          apt-get update
+          apt-get install -y clang-tidy-14 clang-tools-14 python3-pip
+          pip install pyyaml
+          cmake .
```
Comment (mvieth) [inline .github/workflows/clang-tidy.yml:22]: I think you have to tell cmake explicitly to use clang, otherwise it defaults to gcc. This leads to the omp.h-not-found-errors, but might also lead to other errors

Comment (gnawme) [reply .github/workflows/clang-tidy.yml:22]: AFAIK, the `cmake .` compiles `compile_commands.json`, which should contain the correct compiler flags.

Also, this:
[>Clang-tidy](https://developers.redhat.com/blog/2021/04/06/get-started-with-clang-tidy-in-red-hat-enterprise-linux) uses the same front-end libraries as the [Clang C language compiler](https://clang.llvm.org/). However, because it only takes source files as input, you can use clang-tidy for any C or C++ codebase no matter what compiler you are using.

Comment (mvieth) [reply .github/workflows/clang-tidy.yml:17]: > AFAIK, the `cmake .` creates `compile_commands.json`, which should contain the correct compiler flags.

The correct compiler flags for the compiler chosen by cmake, right? So far that was gcc, so `compile_commands.json` contains the compiler flags needed by gcc. I guess clang-tidy needs the compiler flags for clang? That would include [something for omp, probably `-fopenmp`](https://stackoverflow.com/questions/33400462/omp-h-file-not-found-when-compiling-using-clang)

> Also, [this](https://developers.redhat.com/blog/2021/04/06/get-started-with-clang-tidy-in-red-hat-enterprise-linux):
> 
> > Clang-tidy uses the same front-end libraries as the [Clang C language compiler](https://clang.llvm.org/). However, because it only takes source files as input, you can use clang-tidy for any C or C++ codebase no matter what compiler you are using.

I would read that as "You can use clang-tidy even if you normally compile with gcc or another compiler", but in my understanding that doesn't contradict that clang-tidy needs flags for clang (of course this only matters when gcc and clang need different flags, which is the case in PCL, e.g. because of openmp)

BTW you currently don't tell clang-tidy where the `compile_commands.json` is located, or that it should search for the file at all. Does it still use `compile_commands.json`? The logs don't mention anything ...

Comment (gnawme) [reply .github/workflows/clang-tidy.yml]: > BTW you currently don't tell clang-tidy where the `compile_commands.json` is located, or that it should search for the file at all. Does it still use `compile_commands.json`? The logs don't mention anything ...

It's baked into the `run-clang-tidy` script:
```
db_path = 'compile_commands.json'
```

Comment (mvieth) [reply .github/workflows/clang-tidy.yml]: > It's baked into the `run-clang-tidy` script:

Ah, I see, thanks!

Please try telling cmake to use the clang compiler (e.g. like `CC=clang-14 CXX=clang++-14 cmake .`). For me locally at least, that fixed the omp.h-not-found errors.

File: .github/workflows/clang-tidy.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    - uses: actions/checkout@v2
+
+    - name: Run clang-tidy
+      run: |
+          bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)"
+          cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_COMPILER=/usr/bin/clang-14 -DCMAKE_C_COMPILER=/usr/bin/clang-14 .
+          run-clang-tidy -header-filter='.*' -fix
+          git diff > tidy.patch
+          cat tidy.patch
+          [ ! -s tidy.patch ]
```
[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy.yml:21]: > Do we even need to run clang-tidy with `-fix` and then perform these additional steps? As long as the errors are shown, and the script returns with an error code, that's all we want. And `run-clang-tidy` seems to do all that by itself 🙂

Hmm, good point. I'll give that a try.

**Comment by mvieth** [comment]
- the `-fix` flag is probably also not needed?
- specifying `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` here is a good idea. I would then suggest to remove `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` from https://github.com/PointCloudLibrary/pcl/blob/master/CMakeLists.txt#L11 again because that is only needed for clang-tidy
- please enable the remaining (non-default) PCL modules apps, benchmarks, examples, simulation, global_tests. See for example here: https://github.com/PointCloudLibrary/pcl/blob/master/.ci/azure-pipelines/build/ubuntu.yaml . I hope there aren't too many clang-tidy issues in those modules, otherwise it might be better to do that in a separate PR

**Comment by mvieth** [comment]
@gnawme If you rebase this and all checks pass, then this is ready to merge from my side. We can leave enabling the remaining PCL modules for other pull requests

**Comment by mvieth** [comment]
Unfortunately, there are still 4 clang-tidy errors left. Let's fix them in this pull request (not a separate one this time)?

**Review by mvieth — APPROVED** [review]
Nice, thanks!
